### PR TITLE
SSH-Keygen with KDF rounds specified

### LINKED ISSUE #78473: Suggest to Add  ssh-keygen -a (Trials) Option
### Summary

I am trying to achieve a best practise solution with ansible. I have found out that the -a Option from the underlying ssh-keygen is missing.

The -a Option would it allow to improve the security of the generated keys when creating users and their ssh-keys 
[SSH-Keygen](https://linux.die.net/man/1/ssh-keygen)

As the -a Option can help to protect against brute forcing.

The default is 100

### Issue Type

Feature Idea

### Component Name

user

### Additional Information

```yaml (paste below)
- name: Create a 2048-bit SSH key for user jsmith in ~jsmith/.ssh/id_rsa
  ansible.builtin.user:
    name: jsmith
    generate_ssh_key: yes
    ssh_key_bits: 2048
    ssh_key_file: .ssh/id_rsa
    ssh_key_trials: 1000
    ssh_key_passphrase: foo
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by dorpauli**
##### SUMMARY

Suggest to Add ssh-keygen -a (key derivation function rounds) Option based on https://github.com/ansible/ansible/pull/78655. Fixes https://github.com/ansible/ansible/issues/78473.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
- ansible.builtin.user

### CODE REVIEW

File: lib/ansible/modules/user.py
Comment (Akasurde) [inline lib/ansible/modules/user.py]:
```suggestion
            - Optionally specify the number of KDF (key derivation function) rounds used to provide higher resistance to brute-force password cracking.
            - If set to higher values, decrypting the encrypted private key file will take longer.
            - It only takes effect when O(ssh_key_passphrase) is provided because in that case the private key is encrypted.
```

File: test/integration/targets/user/tasks/main.yml
Comment (Akasurde) [inline test/integration/targets/user/tasks/main.yml]: Would it be possible to lower this number, since the integration test will be slower and time-consuming? Ultimately we are just checking if the value of KDF works or not. Just a suggestion, not a blocker.

Comment (dorpauli) [reply test/integration/targets/user/tasks/main.yml]: This is a very good suggestion i think. I change it immediately. Thanks 😃

**Comment by dorpauli** [comment]
@Akasurde Do you know how to get jq properly installed in macos? I tried package and homebrew module, but it seems i cannot use community modules in test context. If there is no solution i will try a solution with grep, cut and awk instead...

**Comment by sivel** [comment]
I wouldn't use `jq`.  Just register the output of the command, then use `from_json` filter, and jinja to access the value you need.

File: test/integration/targets/user/tasks/test_ssh_key_rounds.yml
Comment (mattclay) [inline test/integration/targets/user/tasks/test_ssh_key_rounds.yml]: Use the `from_json` filter to parse the result instead of relying on `jq`.

File: lib/ansible/modules/user.py
Comment (dorpauli) [reply lib/ansible/modules/user.py]: I have changed this part. But because different distributions and versions have different default values, i still included a sentence to mention this. Is this ok, or should i remove defaults completely?

**Comment by dorpauli** [comment]
Do someone know why MacOS fails? It seems that the -a option is useless here. But i cannot see that Darwin class overwrites the ssh_key_gen() method so it should be used. Is -a a proper option in ssh-keygen of mac?

**Comment by dorpauli** [comment]
> Do someone know why MacOS fails? It seems that the -a option is useless here. But i cannot see that Darwin class overwrites the ssh_key_gen() method so it should be used. Is -a a proper option in ssh-keygen of mac?

Or is an ssh-key automatically generated on user creation? So that no ssh-key is generated by module anymore?

**Comment by konstruktoid** [comment]
See the above error messages, https://github.com/ansible/ansible/pull/83468#issuecomment-2177867671, and https://github.com/ansible/ansible/pull/83468/checks?check_run_id=26405481710

**Comment by dorpauli** [comment]
> See the above error messages, [#83468 (comment)](https://github.com/ansible/ansible/pull/83468#issuecomment-2177867671), and https://github.com/ansible/ansible/pull/83468/checks?check_run_id=26405481710

@konstruktoid i already fixed the sanity check locally. Will push it after fixing MacOS. Do you know what could caused the fail in macOS. See my last to comments for this.

**Comment by konstruktoid** [comment]
The only macOS error I see is https://dev.azure.com/ansible/ansible/_build/results?buildId=115391&view=logs&j=12093f7c-b95c-5d17-9f13-4ed61d3451c4&t=5a033008-8dde-5bb0-b993-2fd3ad03089e&l=6159 and that 24 rounds are returned instead of 10. 

I don't own a macOS machine so I can't check the manfile for the macOS defaults or options, and Apple doesn't seem to publish manfiles online.

File: lib/ansible/modules/user.py
Comment (konstruktoid) [inline lib/ansible/modules/user.py:191]: I believe this is incorrect, Ubuntu 24.04 uses 16 rounds by default: https://manpages.ubuntu.com/manpages/noble/man1/ssh-keygen.1.html

Comment (dorpauli) [reply lib/ansible/modules/user.py:191]: Documentation differs from actual test result. That's why i had to adapt the tests.

Comment (konstruktoid) [reply lib/ansible/modules/user.py:191]: TIL.

```
$ python3 -m openssh_key test_id_ed25519 --passphrase secret_passphrase | jq .kdf_options.data.rounds
24
```

**Comment by dorpauli** [comment]
I think my pull request is now ready to be reviewed. While debugging, i noticed two problems that, when combined, caused the test to fail.

1. With regard to the "remove" option, the documentation does not match the behavior of the DarwinUser class for MacOS. From line 2524 (in this PR), you can see that only "force" triggers deletion of the home directory. That is not wrong per se. Nevertheless, "remove" should also trigger deletion. An issue should be opened for this and either the documentation or the method should be adjusted.
2. The "force" parameter does not affect the creation of an SSH key in any way if a passphrase is specified. The documentation does say that this is only overwritten if "force: true" is set. But that does not happen even then. The reason is that the "overwrite" variable created in line 1202 does not affect the query in line 1255. This makes "force" completely pointless here. "overwrite" is used in line 1268 when an SSH key is created without a passphrase. If the behavior is incorrect, an issue should be created for that as well.

@Akasurde @mattclay @konstruktoid Would you please review the code again when you get the chance.

**Comment by NoToComments** [comment]
> The only macOS error I see is https://dev.azure.com/ansible/ansible/_build/results?buildId=115391&view=logs&j=12093f7c-b95c-5d17-9f13-4ed61d3451c4&t=5a033008-8dde-5bb0-b993-2fd3ad03089e&l=6159 and that 24 rounds are returned instead of 10.
> 
> I don't own a macOS machine so I can't check the manfile for the macOS defaults or options, and Apple doesn't seem to publish manfiles online.

Can confirm its 16 rounds as default.
snipping macOs 14 4 manpage

**Comment by dorpauli** [comment]
> > The only macOS error I see is https://dev.azure.com/ansible/ansible/_build/results?buildId=115391&view=logs&j=12093f7c-b95c-5d17-9f13-4ed61d3451c4&t=5a033008-8dde-5bb0-b993-2fd3ad03089e&l=6159 and that 24 rounds are returned instead of 10.
> > I don't own a macOS machine so I can't check the manfile for the macOS defaults or options, and Apple doesn't seem to publish manfiles online.
> 
> Can confirm its 16 rounds as default. snipping macOs 14 4 manpage

Please test it. create an ssh-key with passphrase encrypted. Then use python module as follows:
`$ python3 -m openssh_key ~/.ssh/id_rsa --passphrase secret_passphrase | jq .kdf_options.data.rounds`
If you have MacOS version 14.3 as in the ansible test environment, it should be 24.

**Comment by NoToComments** [comment]
If I simply use `ssh-keygen -N bla `
it creates `/users/blub/.ssh/id_ed25519`

 It is needed to use `ssh-keygen -t rsa -N bla`
 
 And you are right the man page is outdated its 24...
 
Bildschirmfoto 2024-06-19 um 22 28 06

**Comment by dorpauli** [comment]
@Akasurde @mattclay @konstruktoid How long does it usually take for a pull request to be merged? This is my first pull request to ansible.

**Comment by konstruktoid** [comment]
It varies alot @dorpauli, I'm just a contributor so I can't merge anything.

File: lib/ansible/modules/user.py
Comment (bcoca) [inline lib/ansible/modules/user.py]: Normally use  `module.warn()` instead this should be part of argspec, see `required_` functions

File: lib/ansible/modules/user.py
Comment (bcoca) [reply lib/ansible/modules/user.py:191]: ` Default value depends on Platform and Distribution.` is enough, any example will be outdated at one point and require updating.

Comment (dorpauli) [reply lib/ansible/modules/user.py:191]: The same applies to the automated tests. So should i still change this? @bcoca

File: lib/ansible/modules/user.py
Comment (dorpauli) [reply lib/ansible/modules/user.py]: I changed this. Thanks for the hint. What is stillt to do to get this PR merged?

File: lib/ansible/modules/user.py
Comment (bcoca) [inline lib/ansible/modules/user.py]: don't need this

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/user.py
Comment (bcoca) [reply lib/ansible/modules/user.py]: Not much, the main consideration at this point is 'overload' of the user module, many of these options seem to belong on a 'keygen' module instead, bringing this up to core to make a final decision.

**Comment by dorpauli** [comment]
@bcoca What about the change? Has there been a decision in the core team?

**Comment by bcoca** [comment]
Hi!

Thanks very much for your submission to Ansible.  It means a lot to us that you've taken time to contribute.

Unfortunately, we're not sure if we want this feature in the program. We discussed this in corediscussed  and concluded this does not belong in the user module , it is outside it's core competency. Key generation in the user module only has minimal functionality. A far a more complete action  `community.crypto.openssh_keypair` https://docs.ansible.com/ansible/latest/collections/community/crypto/openssh_keypair_module.html exists for those that want to have fine grain controlled on key generation.

However, we're absolutely always up for discussion.
Because this project is very active, we're unlikely to see comments made on closed tickets and we lock them after some time.
If you or anyone else has any further questions, please let us know by using any of the communication methods listed in the page below:

   * https://docs.ansible.com/ansible/latest/community/communication.html

In the future, sometimes starting a discussion on the development list prior to implementing a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!

**Pull request closed without merging**
